File: impl/python/uov/field.py
"""GF(q) arithmetic primitives (q prime)."""

import random
from typing import List, Optional


def gf_inv(a: int, q: int) -> int:
    """Modular inverse of a in GF(q) via Fermat's little theorem."""
    if a % q == 0:
        raise ZeroDivisionError("zero has no inverse in GF(q)")
    return pow(int(a), q - 2, q)


def dot(u: List[int], v: List[int], q: int) -> int:
    return sum(x * y for x, y in zip(u, v)) % q


def mat_mul_vec(M: List[List[int]], v: List[int], q: int) -> List[int]:
    return [dot(row, v, q) for row in M]


def mat_add(A: List[List[int]], B: List[List[int]], q: int) -> List[List[int]]:
    return [[(A[i][j] + B[i][j]) % q for j in range(len(A[0]))] for i in range(len(A))]


def mat_mul(A: List[List[int]], B: List[List[int]], q: int) -> List[List[int]]:
    rows, mid, cols = len(A), len(B), len(B[0])
    return [
        [sum(A[i][k] * B[k][j] for k in range(mid)) % q for j in range(cols)]
        for i in range(rows)
    ]


def gauss_solve(M: List[List[int]], b: List[int], q: int) -> Optional[List[int]]:
    """Solve M·x = b over GF(q). Returns None if M is singular."""
    n = len(b)
    # Augmented matrix [M | b]
    aug = [[M[i][j] for j in range(n)] + [b[i]] for i in range(n)]

    for col in range(n):
        # Find pivot
        pivot = None
        for row in range(col, n):
            if aug[row][col] % q != 0:
                pivot = row
                break
        if pivot is None:
            return None
        aug[col], aug[pivot] = aug[pivot], aug[col]

        inv_p = gf_inv(aug[col][col], q)
        for j in range(col, n + 1):
            aug[col][j] = aug[col][j] * inv_p % q

        for row in range(n):
            if row == col:
                continue
            factor = aug[row][col]
            if factor == 0:
                continue
            for j in range(col, n + 1):
                aug[row][j] = (aug[row][j] - factor * aug[col][j]) % q

    return [aug[i][n] for i in range(n)]


def gf_matinv(M: List[List[int]], q: int) -> Optional[List[List[int]]]:
    """Invert an n×n matrix over GF(q). Returns None if singular."""
    n = len(M)
    # Augmented with identity
    aug = [
        [M[i][j] for j in range(n)] + [1 if i == j else 0 for j in range(n)]
        for i in range(n)
    ]

    for col in range(n):
        pivot = None
        for row in range(col, n):
            if aug[row][col] % q != 0:
                pivot = row
                break
        if pivot is None:
            return None
        aug[col], aug[pivot] = aug[pivot], aug[col]

        inv_p = gf_inv(aug[col][col], q)
        for j in range(2 * n):
            aug[col][j] = aug[col][j] * inv_p % q

        for row in range(n):
            if row == col:
                continue
            factor = aug[row][col]
            if factor == 0:
                continue
            for j in range(2 * n):
                aug[row][j] = (aug[row][j] - factor * aug[col][j]) % q

    return [[aug[i][n + j] for j in range(n)] for i in range(n)]


def gf_random_invertible(n: int, q: int, rng: random.Random) -> List[List[int]]:
    """Sample a random invertible n×n matrix over GF(q)."""
    while True:
        M = [[rng.randrange(q) for _ in range(n)] for _ in range(n)]
        if gf_matinv(M, q) is not None:
            return M


def random_matrix(rows: int, cols: int, q: int, rng: random.Random) -> List[List[int]]:
    return [[rng.randrange(q) for _ in range(cols)] for _ in range(rows)]


def random_vec(n: int, q: int, rng: random.Random) -> List[int]:
    return [rng.randrange(q) for _ in range(n)]

File: impl/rust/src/field.rs
//! GF(q) arithmetic over a prime field (runtime q).

pub type Vec = std::vec::Vec<u64>;
pub type Mat = std::vec::Vec<std::vec::Vec<u64>>;

// ── Scalar operations ─────────────────────────────────────────────────────────

pub fn gf_add(a: u64, b: u64, q: u64) -> u64 { (a + b) % q }
pub fn gf_mul(a: u64, b: u64, q: u64) -> u64 { (a * b) % q }

/// Modular inverse via extended Euclidean algorithm.
pub fn gf_inv(a: u64, q: u64) -> u64 {
    assert!(a % q != 0, "zero has no inverse in GF(q)");
    let (mut old_r, mut r) = (a as i64, q as i64);
    let (mut old_s, mut s) = (1i64, 0i64);
    while r != 0 {
        let quot = old_r / r;
        (old_r, r) = (r, old_r - quot * r);
        (old_s, s) = (s, old_s - quot * s);
    }
    ((old_s % q as i64 + q as i64) as u64) % q
}

pub fn mod_sub(a: u64, b: u64, q: u64) -> u64 {
    ((a as i64 - b as i64).rem_euclid(q as i64)) as u64
}

// ── Vector operations ─────────────────────────────────────────────────────────

pub fn dot(u: &[u64], v: &[u64], q: u64) -> u64 {
    u.iter().zip(v.iter()).fold(0u64, |acc, (&a, &b)| (acc + a * b) % q)
}

pub fn vec_add(u: &[u64], v: &[u64], q: u64) -> Vec {
    u.iter().zip(v.iter()).map(|(&a, &b)| (a + b) % q).collect()
}

pub fn vec_sub(u: &[u64], v: &[u64], q: u64) -> Vec {
    u.iter().zip(v.iter()).map(|(&a, &b)| mod_sub(a, b, q)).collect()
}

// ── Matrix-vector product ─────────────────────────────────────────────────────

pub fn mat_mul_vec(m: &Mat, v: &[u64], q: u64) -> Vec {
    m.iter().map(|row| dot(row, v, q)).collect()
}

// ── Gaussian elimination ──────────────────────────────────────────────────────

/// Solve `m * x = b` over GF(q). Returns `None` if `m` is singular.
pub fn gauss_solve(m: &Mat, b: &[u64], q: u64) -> Option<Vec> {
    let n = b.len();
    // Augmented matrix [m | b]
    let mut aug: Mat = (0..n)
        .map(|i| { let mut row = m[i].clone(); row.push(b[i]); row })
        .collect();

    for col in 0..n {
        let pivot = (col..n).find(|&row| aug[row][col] % q != 0)?;
        aug.swap(col, pivot);

        let inv_p = gf_inv(aug[col][col], q);
        for j in col..=n {
            aug[col][j] = gf_mul(aug[col][j], inv_p, q);
        }

        for row in 0..n {
            if row == col || aug[row][col] == 0 { continue; }
            let factor = aug[row][col];
            for j in col..=n {
                let sub = gf_mul(factor, aug[col][j], q);
                aug[row][j] = mod_sub(aug[row][j], sub, q);
            }
        }
    }

    Some((0..n).map(|i| aug[i][n]).collect())
}

// ── Matrix inverse ────────────────────────────────────────────────────────────

pub fn gf_matinv(m: &Mat, q: u64) -> Option<Mat> {
    let n = m.len();
    let mut aug: Mat = (0..n)
        .map(|i| {
            let mut row = m[i].clone();
            row.extend((0..n).map(|j| if i == j { 1 } else { 0 }));
            row
        })
        .collect();

    for col in 0..n {
        let pivot = (col..n).find(|&row| aug[row][col] % q != 0)?;
        aug.swap(col, pivot);

        let inv_p = gf_inv(aug[col][col], q);
        for j in 0..2*n {
            aug[col][j] = gf_mul(aug[col][j], inv_p, q);
        }

        for row in 0..n {
            if row == col || aug[row][col] == 0 { continue; }
            let factor = aug[row][col];
            for j in 0..2*n {
                let sub = gf_mul(factor, aug[col][j], q);
                aug[row][j] = mod_sub(aug[row][j], sub, q);
            }
        }
    }

    Some((0..n).map(|i| (0..n).map(|j| aug[i][n + j]).collect()).collect())
}

// ── Random helpers ────────────────────────────────────────────────────────────

/// Simple seeded LCG for deterministic random numbers (no external deps).
pub struct Rng { state: u64 }

impl Rng {
    pub fn new(seed: u64) -> Self { Self { state: seed ^ 0x12345678ABCD } }

    pub fn next_u64(&mut self) -> u64 {
        // Splitmix64
        self.state = self.state.wrapping_add(0x9e3779b97f4a7c15);
        let mut z = self.state;
        z = (z ^ (z >> 30)).wrapping_mul(0xbf58476d1ce4e5b9);
        z = (z ^ (z >> 27)).wrapping_mul(0x94d049bb133111eb);
        z ^ (z >> 31)
    }

    pub fn next_field(&mut self, q: u64) -> u64 { self.next_u64() % q }

    pub fn random_vec(&mut self, n: usize, q: u64) -> Vec {
        (0..n).map(|_| self.next_field(q)).collect()
    }

    pub fn random_mat(&mut self, rows: usize, cols: usize, q: u64) -> Mat {
        (0..rows).map(|_| self.random_vec(cols, q)).collect()
    }

    pub fn random_invertible(&mut self, n: usize, q: u64) -> Mat {
        loop {
            let m = self.random_mat(n, n, q);
            if gf_matinv(&m, q).is_some() { return m; }
        }
    }
}
